cli/tests/test_rebase_command.rs

```diff
@@ -631,3 +631,145 @@ fn test_rebase_with_descendants() {
+    create_commit(&test_env, &repo_path, "a", &["base"]);
+    create_commit(&test_env, &repo_path, "b", &["base", "a"]);
+    create_commit(&test_env, &repo_path, "c", &["b"]);
+    let setup_opid = test_env.current_operation_id(&repo_path);
+
+    // Test the setup
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  c
+    ◉    b
+    ├─╮
+    │ ◉  a
+    ├─╯
+    ◉  base
+    ◉
+    "###);
+
+    let (stdout, stderr) =
+        test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "base", "-d", "root()"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Also rebased 4 descendant commits onto parent of rebased commit
+    Working copy now at: vruxwmqv 1fdab507 c | c
+    Parent commit      : royxmykx 4d413a39 b | b
+    Added 0 files, modified 0 files, removed 1 files
+    "###);
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    ◉  base
+    │ @  c
+    │ ◉  b
+    │ ◉  a
+    ├─╯
+    ◉
+    "###);
+
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "base", "-d", "b"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @"");
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    ◉  base
+    │ @  c
+    ├─╯
+    ◉  b
+    ◉  a
+    ◉
+    "###);
+
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "base", "-d", "a"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @"");
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    ◉  base
+    │ @  c
+    │ ◉  b
+    ├─╯
+    ◉  a
+    ◉
+    "###);
+
+    test_env.jj_cmd_ok(&repo_path, &["op", "restore", &setup_opid]);
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "a", "-d", "root()"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Also rebased 2 descendant commits onto parent of rebased commit
+    Working copy now at: vruxwmqv 8d3a8426 c | c
+    Parent commit      : royxmykx 4ea418a3 b | b
+    Added 0 files, modified 0 files, removed 1 files
+    "###);
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    ◉  a
+    │ @  c
+    │ ◉  b
+    │ ◉  base
+    ├─╯
+    ◉
+    "###);
+
+    test_env.jj_cmd_ok(&repo_path, &["op", "restore", &setup_opid]);
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "b", "-d", "root()"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Also rebased 1 descendant commits onto parent of rebased commit
+    Working copy now at: vruxwmqv 62ca8ca6 c | c
+    Parent commit      : zsuskuln 2c5b7858 a | a
+    Added 0 files, modified 0 files, removed 1 files
+    "###);
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    ◉  b
+    │ @  c
+    │ ◉  a
+    │ ◉  base
+    ├─╯
+    ◉
+    "###);
```

Comment:
Just to confirm, some of these test outputs are supposed to change in future?

For example, this will be ?
```
    ◉  b
    │ @  c
    │ ├─╮
    │ │ ◉  a
    │ ├─╯
    │ ◉  base
    ├─╯
    ◉
```

It would be easier to follow if these tests had inline comments about the future changes.

Reply:
I can add some comments. In a few of these cases, I'm not 100% sure what's most convenient for the user though, the way the discussion in #2600 is going, we might just have to preserve the shape regardless.

In this specific case, though, it's pretty clear that the user would want the graph you drew.

Reply:
I think I'll also add a few tests for `rebase -s` and `rebase -b`, possibly in a separate PR.

Reply:
I added them, but it's a good thing I took a second look. See my other comment (that I'm about to write).